main.rb

```diff
@@ -3,17 +3,179 @@
 module Enumerable
   # Task no.1
   def my_each
-    i = 0
-    while self[i]
-      yield(self[i])
-      i += 1
+    if block_given?
+      i = 0
+      while self[i]
+        yield(self[i])
+        i += 1
+      end
+      self
+    else
+      to_enum(:my_each)
     end
-    self
   end
 
   # # TEST FOR #my_each
-  # [1,2,3,4].my_each do |x|
+  # [1, 2, 3, 4].my_each do |x|
   #   puts "#{x}"
   # end
 
+  # Task no.2
+  def my_each_with_index
+    if block_given?
+      i = 0
+      while self[i]
+        yield(self[i], i)
+        i += 1
+      end
+      self
+    else
+      to_enum(:my_each_with_index)
+    end
+  end
+
+  # TEST FOR #my_each_with_index
+  # [1, 2, 3, 4].my_each_with_index do |x, i|
+  #   puts "#{i}. #{x}"
+  # end
+
+  # Task no.3
+  def my_select
+    if block_given?
+      new_array = []
+      my_each do |i|
+        new_array << i if yield(i)
+      end
+      new_array
+    else
+      to_enum(:my_select)
+    end
+  end
+
+  # TEST FOR my_select
+  # num = [1, 2, 3, 4, 5, 6, 7, 8]
+  # p num.my_select { |x| x.even? }
+  # p num
+
+  # Task no.4
+  def my_all?
+    if block_given?
+      my_each do |i|
+        return false if yield(self[i]) == false || yield(self[i]).nil?
+      end
+      true
+    else
+      to_enum(:my_all?)
+    end
+  end
+
+  # TEST FOR my_all?
+  # num = [1, 2, 3, 4, 5]
+  # p num.my_all? { |x| x > 6}
+
+  # Task no.5
+  def my_any?
+    if block_given?
+      my_each do |f|
+        return true if yield(f)
```

Comment:
- [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern